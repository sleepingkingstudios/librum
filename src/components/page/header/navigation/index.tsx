import * as React from 'react';
import { map } from 'lodash';

import { joinClassNames } from '@utils/react-utils';
import {
  Dropdown,
  DropdownProps,
} from './dropdown';
import {
  NavigationItem,
  NavigationItemProps,
} from './item';
import { useThemeStyles } from '@themes';

export type NavigationProps = Array<NavigationItemProps | DropdownProps>;

interface PageNavigationProps {
  className?: string;
  navigation: NavigationProps;
}

const defaultListClassName = 'flex flex-col sm:flex-row';

export const PageNavigation = (
  { className, navigation }: PageNavigationProps
): JSX.Element => {
  const joinedClassNames = joinClassNames(
    'relative z-10',
    useThemeStyles('navigation'),
    className,
  );
  const joinedListNames = joinClassNames(
    defaultListClassName,
    useThemeStyles('navigation-list'),
  );
  const [open, setOpen] = React.useState<string | null>(null);

  return (
    <nav className={joinedClassNames}>
      <ul className={joinedListNames}>
        {
          map(
            navigation,
            (item: DropdownProps | NavigationItemProps): JSX.Element => {
              const { label } = item;

              if('items' in item) {
                return (
                  <Dropdown
                    key={label}
                    open={open}
                    setOpen={setOpen}
                    {...item}
                  />
                );
              }

              return (<NavigationItem key={label} {...item} />);
            },
          )
        }
      </ul>
    </nav>
  );
};
