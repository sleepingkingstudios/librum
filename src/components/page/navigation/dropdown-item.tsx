import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

import type { NavigationItemProps } from './item';

export const DropdownItem = ({ label, url }: NavigationItemProps): JSX.Element => {
  const { pathname } = useLocation();

  if (pathname === url) {
    return (
      <li className="dropdown-item">
        <p className="dropdown-item-link-disabled text-muted">{ label }</p>
      </li>
    );
  } else {
    return (
      <li className="dropdown-item">
        <Link className="dropdown-item-link" to={url}>
          { label }
        </Link>
      </li>
    );
  }
};
