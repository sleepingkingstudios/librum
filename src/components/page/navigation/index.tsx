import * as React from 'react';
import { map } from 'lodash';

import { joinClassNames } from '@utils/react-utils';
import {
  Dropdown,
  DropdownProps,
} from './dropdown';
import {
  NavigationItem,
  NavigationItemProps,
} from './item';

export type NavigationProps = Array<NavigationItemProps | DropdownProps>;

interface PageNavigationProps {
  className?: string;
  navigation: NavigationProps;
}

export const PageNavigation = (
  { className, navigation }: PageNavigationProps
): JSX.Element => {
  const joinedClassNames = joinClassNames(
    'page-navigation relative z-10',
    className,
  );
  const [open, setOpen] = React.useState<string | null>(null);

  return (
    <nav className={joinedClassNames}>
      <ul className="flex flex-col sm:flex-row">
        {
          map(
            navigation,
            (item: DropdownProps | NavigationItemProps): JSX.Element => {
              const { label } = item;

              if('items' in item) {
                return (
                  <Dropdown
                    key={label}
                    open={open}
                    setOpen={setOpen}
                    {...item}
                  />
                );
              }

              return (<NavigationItem key={label} {...item} />);
            },
          )
        }
      </ul>
    </nav>
  );
};
