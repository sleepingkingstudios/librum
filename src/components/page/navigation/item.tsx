import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

export interface NavigationItemProps {
  label: string;
  url: string;
}

export const NavigationItem = ({ label, url }: NavigationItemProps): JSX.Element => {
  const { pathname } = useLocation();

  if (pathname === url) {
    return (
      <li className="navigation-item">
        <p className="navigation-item-link-disabled text-muted">{ label }</p>
      </li>
    );
  } else {
    return (
      <li className="navigation-item">
        <Link className="navigation-item-link" to={url}>{ label }</Link>
      </li>
    );
  }
};
