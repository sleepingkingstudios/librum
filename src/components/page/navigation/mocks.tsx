import * as React from 'react';

import type { DropdownProps } from './dropdown';
import type { NavigationItemProps } from './item';

const renderNavigationItem = (
  item: DropdownProps | NavigationItemProps
): JSX.Element => {
  const { label } = item;

  if ('items' in item) {
    return (
      <li key={label}>
        { label }
        <ul>
          { item.items.map(renderNavigationItem) }
        </ul>
      </li>
    );
  }

  const { url } = item;

  return (
    <li aria-label="navigation-item" key={label}>{ `${label} @ ${url}` }</li>
  );
};

export const PageNavigation = ({
  navigation,
}: {
  navigation: Array<DropdownProps | NavigationItemProps>,
}): JSX.Element => {
  if (!navigation) { return null; }

  if (navigation.length === 0) { return null; }

  return (
    <ul aria-label="navigation">
      {
        navigation.map(renderNavigationItem)
      }
    </ul>
  );
};
