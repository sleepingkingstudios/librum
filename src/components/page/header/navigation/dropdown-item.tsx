import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

import { NavigationItemProps } from './item';
import { useThemeStyles } from '@themes';

export const DropdownItem = ({ label, url }: NavigationItemProps): JSX.Element => {
  const className = useThemeStyles('navigation-dropdown-item');
  const linkClassName = useThemeStyles('navigation-dropdown-link');
  const disabledClassName = useThemeStyles('navigation-dropdown-link-disabled');
  const { pathname } = useLocation();

  if (pathname === url) {
    return (
      <li className={className}>
        <p className={disabledClassName}>{ label }</p>
      </li>
    );
  } else {
    return (
      <li className={className}>
        <Link className={linkClassName} to={url}>
          { label }
        </Link>
      </li>
    );
  }
};
