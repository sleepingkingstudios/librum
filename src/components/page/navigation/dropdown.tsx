import * as React from 'react';
import { map } from 'lodash';
import { FontAwesomeIcon as Icon } from '@fortawesome/react-fontawesome';
import {
  faCaretDown,
  faCaretUp,
} from '@fortawesome/free-solid-svg-icons';

import { DropdownItem } from './dropdown-item';
import { NavigationItemProps } from './item';

export interface DropdownProps {
  label: string;
  items: Array<NavigationItemProps>;
}

export interface DropdownElementProps extends DropdownProps {
  open: string;
  setOpen: React.Dispatch<React.SetStateAction<string | null>>;
}

interface renderDropdownItemsProps {
  items: Array<NavigationItemProps>;
  open: boolean;
}

const renderDropdownItem = (item: NavigationItemProps): JSX.Element => {
  const { label } = item;

  return (<DropdownItem key={label} {...item} />);
};

const renderDropdownItems = (
  { items, open }: renderDropdownItemsProps
): JSX.Element => {
  if (!open) { return null; }

  return (
    <ul>
      { map(items, renderDropdownItem) }
    </ul>
  )
}

export const Dropdown = (
  {
    label,
    items,
    open,
    setOpen,
  }: DropdownElementProps
): JSX.Element => {
  const isOpen = label === open;
  const openDropdown = () => setOpen(label);
  const closeDropdown = () => setOpen(null);

  return (
    <li className="dropdown-container">
      <div className="dropdown">
        <button
          onClick={isOpen ? closeDropdown : openDropdown}
          className="dropdown-label"
        >
          { label }
          <Icon
            icon={isOpen ? faCaretDown : faCaretUp}
            size='xs'
            className='ml-2'
          />
        </button>

        { renderDropdownItems({ items, open: isOpen }) }
      </div>
    </li>
  );
}
