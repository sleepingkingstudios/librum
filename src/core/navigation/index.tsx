import * as React from 'react';

import { PageNavigation } from '@components/page/navigation';

const navigation = [
  {
    label: 'Home',
    url: '/',
  },
  {
    label: 'Publishers',
    url: '/publishers',
  },
  {
    label: 'Game Systems',
    url: '/game_systems',
  },
  {
    label: 'Sources',
    items: [
      {
        label: 'Books',
        url: '/sources/books',
      },
      {
        label: 'Websites',
        url: '/sources/websites',
      },
    ],
  },
  {
    label: 'Admin',
    items: [
      {
        label: 'Users',
        url: '/admin/users',
      },
    ],
  },
];

export const CoreNavigation = () => (
  <PageNavigation navigation={navigation} />
);
