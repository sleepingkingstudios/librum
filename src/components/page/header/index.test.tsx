import * as React from 'react';

import { screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { PageHeader } from './index';
import { actions } from '@session';
import type { User } from '@session';
import { render } from '@test-helpers/rendering';
import { createStore } from '@test-helpers/store';
import { PageNavigation } from '../navigation';

describe('<PageHeader>', () => {
  it('should display the default title', () => {
    const defaultTitle = 'Librum';

    render(<PageHeader />, { store: true });

    const text = screen.getByText(defaultTitle);

    expect(text).toBeVisible();
  });

  it('should display the default subtitle', () => {
    const defaultSubtitle = 'Campaign Companion';

    render(<PageHeader />, { store: true });

    const text = screen.getByText(defaultSubtitle);

    expect(text).toBeVisible();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <PageHeader />,
      { store: true },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with navigation: component', () => {
    const navigation = [
      {
        label: 'Home',
        url: '/',
      },
      {
        label: 'Launch Sites',
        url: '/launch-sites',
      },
      {
        label: 'Rockets',
        items: [
          {
            label: 'Engines',
            url: '/rockets/engines',
          },
          {
            label: 'Fuel Tanks',
            url: '/rockets/fuel-tanks',
          },
        ],
      },
      {
        label: 'Administration',
        items: [
          {
            label: 'Strategies',
            url: '/administration/strategies',
          },
        ],
      },
    ];
    const Navigation = () => (<PageNavigation navigation={navigation} />);

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PageHeader navigation={<Navigation />} />,
        {
          router: true,
          store: true,
        },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with navigation: value', () => {
    const navigation = [
      {
        label: 'Home',
        url: '/',
      },
      {
        label: 'Launch Sites',
        url: '/launch-sites',
      },
      {
        label: 'Rockets',
        items: [
          {
            label: 'Engines',
            url: '/rockets/engines',
          },
          {
            label: 'Fuel Tanks',
            url: '/rockets/fuel-tanks',
          },
        ],
      },
      {
        label: 'Administration',
        items: [
          {
            label: 'Strategies',
            url: '/administration/strategies',
          },
        ],
      },
    ];

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PageHeader navigation={navigation} />,
        {
          router: true,
          store: true,
        },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with subtitle: null', () => {
    it('should not display the subtitle', () => {
      const defaultTitle = 'Librum';

      render(<PageHeader subtitle={null} />, { store: true });

      const text = screen.getByText(defaultTitle);

      expect(text).toBeVisible();
      expect(text).toHaveTextContent(/^Librum$/);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PageHeader />,
        { store: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with subtitle: value', () => {
    const subtitle = 'Example Subtitle';

    it('should display the configured title', () => {
      render(<PageHeader subtitle={subtitle} />, { store: true });

      const text = screen.getByText(subtitle);

      expect(text).toBeVisible();
    });
  });

  describe('with title: value', () => {
    const title = 'Example Title';

    it('should display the configured title', () => {
      render(<PageHeader title={title} />, { store: true });

      const text = screen.getByText(title);

      expect(text).toBeVisible();
    });
  });

  describe('when the session is authenticated', () => {
    const user: User = {
      email: 'alan.bradley@example.com',
      id: '00000000-0000-0000-0000-000000000000',
      role: 'user',
      slug: 'alan-bradley',
      username: 'Alan Bradley',
    };
    const token = '12345';
    const { create } = actions;

    it('should match the snapshot', () => {
      const { dispatch, store } = createStore();
      dispatch(create({ token, user }));

      const { asFragment } = render(
        <PageHeader />,
        {
          router: true,
          store,
        },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
