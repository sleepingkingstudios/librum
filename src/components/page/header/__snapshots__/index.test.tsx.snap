// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageHeader> should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="mb-3 text-center sm:text-left"
  >
    <span
      class="page-title"
    >
      Librum
    </span>
    <br
      class="block sm:hidden"
    />
    <small
      class="page-subtitle m-0 sm:ml-2"
    >
      Campaign Companion
    </small>
    <hr
      class="hr-fancy my-1"
    />
  </header>
</DocumentFragment>
`;

exports[`<PageHeader> when the session is authenticated should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="mb-3 text-center sm:text-left"
  >
    <span
      class="page-title"
    >
      Librum
    </span>
    <br
      class="block sm:hidden"
    />
    <small
      class="page-subtitle m-0 sm:ml-2"
    >
      Campaign Companion
    </small>
    <hr
      class="hr-fancy my-1"
    />
    <div
      class="text-left mt-2"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-user mr-1"
        data-icon="user"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
          fill="currentColor"
        />
      </svg>
      <span
        class="hidden sm:inline"
      >
        You are currently logged in as 
      </span>
      <a
        href="/user"
      >
        Alan Bradley
      </a>
      <span
        class="float-right link-danger"
      >
        <button>
          Log Out
        </button>
      </span>
    </div>
  </header>
</DocumentFragment>
`;

exports[`<PageHeader> with navigation: component should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="mb-3 text-center sm:text-left"
  >
    <span
      class="page-title"
    >
      Librum
    </span>
    <br
      class="block sm:hidden"
    />
    <small
      class="page-subtitle m-0 sm:ml-2"
    >
      Campaign Companion
    </small>
    <nav
      class="page-navigation relative z-10"
    >
      <ul
        class="flex flex-col sm:flex-row"
      >
        <li
          class="navigation-item"
        >
          <p
            class="navigation-item-link-disabled text-muted"
          >
            Home
          </p>
        </li>
        <li
          class="navigation-item"
        >
          <a
            class="navigation-item-link"
            href="/launch-sites"
          >
            Launch Sites
          </a>
        </li>
        <li
          class="dropdown-container"
        >
          <div
            class="dropdown"
          >
            <button
              class="dropdown-label"
            >
              Rockets
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-up fa-xs ml-2"
                data-icon="caret-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </li>
        <li
          class="dropdown-container"
        >
          <div
            class="dropdown"
          >
            <button
              class="dropdown-label"
            >
              Administration
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-up fa-xs ml-2"
                data-icon="caret-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </li>
      </ul>
    </nav>
    <hr
      class="hr-fancy my-1"
    />
  </header>
</DocumentFragment>
`;

exports[`<PageHeader> with navigation: value should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="mb-3 text-center sm:text-left"
  >
    <span
      class="page-title"
    >
      Librum
    </span>
    <br
      class="block sm:hidden"
    />
    <small
      class="page-subtitle m-0 sm:ml-2"
    >
      Campaign Companion
    </small>
    <hr
      class="hr-fancy my-1 block sm:hidden"
    />
    <nav
      class="page-navigation relative z-10"
    >
      <ul
        class="flex flex-col sm:flex-row"
      >
        <li
          class="navigation-item"
        >
          <p
            class="navigation-item-link-disabled text-muted"
          >
            Home
          </p>
        </li>
        <li
          class="navigation-item"
        >
          <a
            class="navigation-item-link"
            href="/launch-sites"
          >
            Launch Sites
          </a>
        </li>
        <li
          class="dropdown-container"
        >
          <div
            class="dropdown"
          >
            <button
              class="dropdown-label"
            >
              Rockets
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-up fa-xs ml-2"
                data-icon="caret-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </li>
        <li
          class="dropdown-container"
        >
          <div
            class="dropdown"
          >
            <button
              class="dropdown-label"
            >
              Administration
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-up fa-xs ml-2"
                data-icon="caret-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </li>
      </ul>
    </nav>
    <hr
      class="hr-fancy my-1"
    />
  </header>
</DocumentFragment>
`;

exports[`<PageHeader> with subtitle: null should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="mb-3 text-center sm:text-left"
  >
    <span
      class="page-title"
    >
      Librum
    </span>
    <br
      class="block sm:hidden"
    />
    <small
      class="page-subtitle m-0 sm:ml-2"
    >
      Campaign Companion
    </small>
    <hr
      class="hr-fancy my-1"
    />
  </header>
</DocumentFragment>
`;
